[WIP] ui trader

askmike opened the pull request:
Before we can add live trading (tradebot) to the UI, we need to do the following:

- [x] register at what prices and times orders where executed at.
  - [x] poloniex
  - [x] kraken
  - [x] bitstamp
  - [x] bitfinex
  - [x] GDAX
  - [x] BTCC (will push this to next version, see #806)
  - [x] BTC-E
- [x] hook into all risk & profit calculation logic (currently inside papertrader)
  - [x] pull performance/risk calculation logic from paperTrader into new plugin
  - [x] broadcast paper trader trades as "trade"
  - [x] broadcast real trader trades as "trade"
- [x] expose via web UI
- [x] add web UI control panel where users can set API keys

askmike committed:
add getOrder to portfolioManager & poloniex

generalectric commented:
> broadcast those as "trades"

Excited about this! Two questions though... How do I access indicator output via api.. Last I checked it simply isn't possible unless I'm blind... I mean can the ui parse indicator output? Am I blind? I want to contribute but I'm kinda road blocked.. Well not blocked I just find it easier to make forward progress tackling easier tasks. And number two... I saw another member created multi time frame ability... And tied that into strategies. I'm curious if ANYONE here actually makes decisions based on one time frame. Me personally I review 5 15 30 min,  2 and 4 hour candles before I make a decision.. I would think my strategies should do the same!  Are there any plans for the ability to tie different indicators to different candle sizes? I believe this work has already been done... Could we get this implemented?

askmike commented:
To answer your questions I have to quickly go over Gekko's architecture. There is a document describing it [here](https://gekko.wizb.it/docs/internals/architecture.html), but it's a but out of date. Very simply put this is how Gekko works:

![screen shot 2017-05-06 at 11 56 15](https://cloud.githubusercontent.com/assets/969743/25771789/4d23c150-3253-11e7-889c-a81d482e1591.png)

This PR is to break the last part (paper trader) up into two different parts (paper trader & performance analyzer) this way we can use exactly the same code to calculate statistics, results, etc. for both the paper trader and real trader:

![screen shot 2017-05-06 at 11 57 44](https://cloud.githubusercontent.com/assets/969743/25771797/95d4340c-3253-11e7-9afc-fc92a7f9fe02.png)

This will be necesarry to support live trading in the UI, since we need to display these statistics for the live trader just as the paper trader.

-------

> How do I access indicator output via api.. Last I checked it simply isn't possible unless I'm blind... I mean can the ui parse indicator output?

Those are not exposed yet, I have plans to do that (see #690). This is a completely separate issue as it will involve changing some code in the strategy logic (the second last box in the first picture - not the strats, but the wrapper code that registers indicators, updates them every candle, etc).

The live trader has higher priority for me.

> I want to contribute but I'm kinda road blocked..

If you want to work on #690 that would be very hard (if you are not familiar with the codebase) since you have to touch the deepest/oldest parts of the codebase as well as the networking/UI logic in order to display the results. The gist is that gekko runs code on every candle (when it executes your `update` and `check` methods) [here](https://github.com/askmike/gekko/blob/stable/plugins/tradingAdvisor/baseTradingMethod.js#L211). This is where we need to take indicator output and push it through [to the parent process](https://github.com/askmike/gekko/blob/stable/core/cp.js). I am planning to work on that myself, but I want to do this first.

> I saw another member created multi time frame ability... And tied that into strategies. I'm curious if ANYONE here actually makes decisions based on one time frame. Me personally I review 5 15 30 min, 2 and 4 hour candles before I make a decision.. I would think my strategies should do the same! Are there any plans for the ability to tie different indicators to different candle sizes? I believe this work has already been done... Could we get this implemented?

That user wanted this to be part of Gekko but I ended up not adding it as is because:

The aim of Gekko is to simplify trading strategies, but the required extra config (for everyone, even people who run simple strategies over a single timeframe) made it very complex to understand how to create a strategy for X timeframes and where to put code that was run under what timeframe. I am very open to a PR or issue discussing integrating that.

That said I think this would be a better approach:

Gekko has code that runs indicators based on candles, but now that is tighly coupled to the strat wrapping code. If we can decouple that as a simple class like so: `put candle in --> [  x   ] --> indicator result comes out`. Once we have those we can very simply expose them to the strategies (in combination with the candleBatcher, a class that takes 1m candles and spits out Xm candles).

-----

**Can we please keep discussion on topic, eg. let's discuss the live trader here and other thing in the corresponding issue.**

askmike committed:
allow for events that can be broadcasted by a number of plugins

askmike committed:
after figuring out which plugin will be emitting, actually use it

askmike committed:
pull performance calcs out of paper trader into performanceAnalyzer plugin

askmike committed:
pull performance calcs out of paper trader into performanceAnalyzer plugin

askmike committed:
configure performanceAnalyzer plugin in UI

askmike committed:
configure performanceAnalyzer plugin in UI

askmike committed:
minor bump

askmike committed:
calculate and relay live trade information (for polo)

askmike committed:
relay real trades & portfolio to performanceAnalyzer

askmike committed:
catch starting new trade while the other one is still running

askmike committed:
properly relay portfolio from trader => performanceAnalyzer

askmike committed:
handle order filled while trying to cancel

askmike committed:
remove deprecated spies

askmike committed:
fix live syncing roundtrips

askmike committed:
add tradebot to UI

askmike commented:
live trading with the UI works for poloniex now:

<img width="1150" alt="screen shot 2017-06-07 at 00 42 47" src="https://user-images.githubusercontent.com/969743/26856214-5df73540-4b1a-11e7-8974-c96690e272dc.png">

but keys are required to be set in `/api-keys.js` (use [this](https://github.com/askmike/gekko/blob/feature/ui-trader/api-keys-sample.js) as a base)

long-einser commented:
what do I need to change at the simulation settings(asset, currency, fee,..) if I would like to trade live? or did I miss something? (i edited the code to support live trading ( https://github.com/askmike/gekko/commit/e515a39a5477eabc1a6eee9ff73e26253278efe9 ))could someone post an example pls?  ty for help&time

askmike commented:
Those commits are part of a new feature which is not completed yet, it
barely works at this stage (limited to poloniex, no edge case/error
handling, logic needs to be validated, etc). I would highly advice not
running it since this is untested and unstable code - as you might lose
money on bugs, but if you do you can just download that branch instead.

On Wed, Jun 14, 2017 at 11:18 AM, long-einser <notifications@github.com>
wrote:

> what do I need to change at the simulation settings(asset, currency,
> fee,..) if I would like to trade live? or did I miss something? (i edited
> the code to support live trading (e515a39
> <https://github.com/askmike/gekko/commit/e515a39a5477eabc1a6eee9ff73e26253278efe9))could>
> someone post an example pls? ty for help&time
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/askmike/gekko/pull/737#issuecomment-308388802>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AA7MDxYTnnk0NVkAUJrYbI7WgUsOedLfks5sD7NbgaJpZM4NRYYN>
> .
>



-- 
PGP key at keybase.io/mikevanrossum
<https://keybase.io/mikevanrossum/key.asc>

long-einser commented:
Thank you for your fast reply. 
I installd and tried the branch out for some days. 
Problems I have now are.. 
I cant make backtests ( all calculations( show up as nan or nan%), same histroy backtested with branch stable it works.
I still have no clue what asset=1 does? I could imagine it but not 100% sure (currency choice?) 
No trades in 2days.

have a nice weekend

askmike commented:
That branch is in development and all features are probably broken, it's not meant to be run right now (very unstable). Whenever I have free time I will work on that branch and improve it. Once it is finished it will be part of the develop branch and can be tested (with maybe just a few small bugs).

> I still have no clue what asset=1 does?

This is a backtest setting (has no effect on real trading). When you run a backtest you are simulating with virtual money, asset=1 means that your virtual balance includes 1 asset (so bitcoin if you are simulating over a new branch). @long-einser please open a new issue if you have further questions. Your comments have nothing to do with the implementation of this issue.

askmike committed:
remove DAO markets from kraken

askmike committed:
retry failed order on polo

askmike committed:
number of fixes with kraken & implement getOrder

askmike committed:
retry buy/sell on bitstamp

askmike committed:
prepare bitstamp trader for the UI (implement getOrder)

askmike committed:
document bitstamps getOrder return behaviour

askmike committed:
improve bitfinex portfolio fetching

askmike committed:
implement getOrder for bitfinex

askmike committed:
retry all critical trade methods at bitfinex

askmike committed:
update btc-e dep, implement getOrder, fix #526, fix #776 & fix #780

askmike committed:
update GDAX dep, minor fixes, implment getOrder, fix #628

askmike committed:
add apiKeyManager service

askmike committed:
plug broadcast into apiKeyManager

askmike committed:
when starting real trader, pull keys out of apiKeyManager

askmike committed:
add config page to UI

askmike committed:
manage API keys in the UI!

askmike committed:
add remove api key buttons

askmike commented:
You can manage your API keys through the gekko UI!

It looks like this:

<img width="911" alt="screen shot 2017-07-06 at 18 51 21" src="https://user-images.githubusercontent.com/969743/27925308-350ce91c-627c-11e7-9ce1-ed819e1e282e.png">

-----

Only thing left to do before we can merge is:

- When you start a trading bot, check whether the API keys are available, if not route to config page.
- Test!

askmike committed:
only warn for overwriting if keys already exist

askmike committed:
force store init in case the WS is down

askmike committed:
fix btce cancel, fix #780

Ali1 commented:
@askmike Just to comment about the api-key configuration design... looks like you are developing it so that there is only 1 set of keys per exchange. Other than the fact that some individuals may want to trade on multiple accounts, Bitstamp have an amazing feature were you can set up sub-accounts. They have created this feature specifically with automated traders in mind. Each sub accounts has its own real-currency and cryptocurrency balances and separate API keys. It's great for traders who want to trade multiple cryptocurrencies or run different strategies simultaneously. 

I hope more exchanges also develop such features in the future.

But my point is if you are moving towards stored api-keys, it would make sense for them to support multiple sets of keys for each exchange - if that isn't the case already.

nourspace committed:
Fix undefined errors in gdax.js

- Changed error to err
- Unified errors checking and displaying
  `err || data.hasOwnProperty('message')`
- Added missing bind to the newly added result callbacks
- Added missing args to retry in sell

askmike committed:
enhance trade problem logging, cleanup debug stuff

askmike commented:
@ali1 good point, I will later update this system to be more flexible. But for now I just want to roll this out to start getting some feedback and bug reports.

I was thinking about improving this by allowing users to "name" a set of keys and also set trading limits tied to this name.

> But my point is if you are moving towards stored api-keys, it would make sense for them to support multiple sets of keys for each exchange - if that isn't the case already.

You are very right, having to later migrate this might be more work. But having multiple possible keysets per exchange also requires a lot of different updates: users need to specifically select a keypair "name" when starting a gekko, and if you are already running a live gekko on the same exchange (using a different keypair "name") we need to be able to understand whether that is an API key of the same account (something Gekko will not support), or a different account (or subaccount). The UI-trader has been in development long enough, just want to get a feature our first :)

askmike committed:
check whether the user has API keys configured of the exchange

askmike committed:
only allow for tradable exchanges to be selected when selecting tradebot

askmike committed:
gekko v0.5.0 & ui v0.2.0

askmike commented:
This is what it looks like to run a live trader:

<img width="1361" alt="screen shot 2017-07-23 at 20 06 11" src="https://user-images.githubusercontent.com/969743/28502417-050d4434-6fea-11e7-91a6-7cc813e4219c.png">

askmike committed:
recompile ui

CyborgDroid commented:
> Just to comment about the api-key configuration design... looks like you are developing it so that there is only 1 set of keys per exchange. Other than the fact that some individuals may want to trade on multiple accounts, Bitstamp have an amazing feature were you can set up sub-accounts. They have created this feature specifically with automated traders in mind. Each sub accounts has its own real-currency and cryptocurrency balances and separate API keys. It's great for traders who want to trade multiple cryptocurrencies or run different strategies simultaneously.

If you have multiple accounts trading at once you would need multiple strategies and settings running against those. This makes it a bit more complicated. But if we'll have multiple separate strategies, it would be great to add in there the ability to create a strategy (on a single account) that can run strategies with different candle sizes so you can compare them before signaling a buy/sell. For example: If strategy1 with 1hr candle size says SELL check that Strategy2 with 5min candle says sell as well, wait until it does before selling. Maybe its already possible, but doesn't seem so in the UI. 

At that point you would be a step away from UI-ing the actual custom strategies with a logic builder since you would have a list of configured strategies.
